### PR TITLE
Add ElasticSearch versions 2.X and 5 Output.

### PR DESCRIPTION
**Pull request by amandahla**
### Required for all PRs:
- [ ] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [X ] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [X ] README.md updated (if adding a new plugin)

### CODE REVIEW

File: plugins/outputs/elasticsearch/elasticsearch_test.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+func (a *Elasticsearch) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+
+	for _, metric := range metrics {
+
+		m := make(map[string]interface{})
+		//m["created"] = metric.UnixNano() / 1000000
+		m["created"] = time.Now()
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: this should be the metric's timestamp. I believe metric.Time()

File: plugins/outputs/elasticsearch/elasticsearch_test.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+		if host, ok := metric.Tags()["host"]; ok {
+			m["host"] = host
+		} else {
+			host, err := os.Hostname()
+			if err != nil {
+				panic(err)
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: don't panic, just log it and move on

File: plugins/outputs/elasticsearch/elasticsearch_test.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+			}
+			m["host"] = host
+		}
+
+		// Earlier versions of EL doesnt accept '.' in field name
+		if len(a.Separator) > 1 {
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: can you explain what this is? is this an undocumented config option?

File: plugins/outputs/elasticsearch/elasticsearch_test.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+			return fmt.Errorf("FAILED Separator exceed one character : %s\n", a.Separator)
+		}
+
+		for key, value := range metric.Tags() {
+			if key != "host" {
+				if strings.HasPrefix(a.Version, "2.") {
+					m[strings.Replace(key, ".", a.Separator, -1)] = value
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: Are you just removing all "." here? I think there's a better way to do this. You shouldn't assume that a "." is always going to be a metric separator. You should setup a "sanitization" regex that can clean up bad characters. See the graphite output for an example.

File: plugins/outputs/elasticsearch/elasticsearch_test.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+				} else {
+					m[key] = value
+				}
+			}
+		}
+
+		for key, value := range metric.Fields() {
+			if strings.HasPrefix(a.Version, "2.") {
+				m[strings.Replace(key, ".", a.Separator, -1)] = value
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: same as above comment

File: plugins/outputs/elasticsearch/elasticsearch_test.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+			} else {
+				m[key] = value
+			}
+		}
+
+		_, errMessage := a.Client.Index().
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: what is the return value being ignored here?

File: plugins/outputs/elasticsearch/elasticsearch_test.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+			Index(a.IndexName).
+			Type(metric.Name()).
+			BodyJson(m).
+			Do()
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: can this panic? is there a non-panicking version of this function call available in the client library?

File: plugins/outputs/elasticsearch/elasticsearch_test.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+func (a *Elasticsearch) SampleConfig() string {
+	return sampleConfig
+}
+
+func (a *Elasticsearch) Description() string {
+	return "Configuration for Elasticsearch to send metrics to."
+}
+
+func buildMetrics(m telegraf.Metric) (map[string]Point, error) {
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: is this being used? why is this here?

File: plugins/outputs/elasticsearch/elasticsearch.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+		if err := p.setValue(v); err != nil {
+			return ms, fmt.Errorf("unable to extract value from Fields, %s", err.Error())
+		}
+		p[0] = float64(m.Time().Unix())
+		ms[k] = p
+	}
+	return ms, nil
+}
+
+func (p *Point) setValue(v interface{}) error {
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch.go]: this isn't being used?

File: plugins/outputs/elasticsearch/elasticsearch.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+type TimeSeries struct {
+	Series []*Metric `json:"series"`
+}
+
+type Metric struct {
+	Metric string   `json:"metric"`
+	Points [1]Point `json:"metrics"`
+}
+
+type Point [2]float64
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch.go]: not used?

**Comment by sparrc** [comment]
since elasticsearch accepts JSON, please write better unit tests that don't only use docker containers.

I'd like to see tests that verify the metrics that go in vs. the JSON that is generated by this plugin, so that it's clear what the elasticsearch metrics look like.

File: plugins/outputs/elasticsearch/elasticsearch_test.go (hunk 3 above)
Comment (amandahla) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: Hello. It's a limitation from earlier versions. Its better explained here:
https://www.elastic.co/guide/en/elasticsearch/reference/current/dots-in-names.html

File: plugins/outputs/elasticsearch/elasticsearch_test.go (hunk 4 above)
Comment (amandahla) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: Yes, I remove all dots from key because this will be used as a field name. I just do that if it's some 2.X version. Others separator its fine, just the '.' its a problem.

Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: so why make that a config option that you're calling "Separator"? Why not just sanitize all strings before you marshal them to json?

File: plugins/outputs/elasticsearch/elasticsearch_test.go (hunk 6 above)
Comment (amandahla) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: I didnt find very helpful since I can use the errMessage to show if something did wrong.

Any ideas?

The return is:
type IndexResponse struct {
	// TODO _shards { total, failed, successful }
	Index   string `json:"_index"`
	Type    string `json:"_type"`
	Id      string `json:"_id"`
	Version int    `json:"_version"`
	Created bool   `json:"created"`
}

File: plugins/outputs/elasticsearch/elasticsearch_test.go (hunk 2 above)
Comment (amandahla) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: Changed in commit:
https://github.com/influxdata/telegraf/pull/1875/commits/862f728635110c159444d2a88bdf3cdf9ae47edc

File: plugins/outputs/elasticsearch/elasticsearch_test.go (hunk 7 above)
Comment (amandahla) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: I dont think so. The example code (http://olivere.github.io/elastic/) and Cadvisor (https://github.com/google/cadvisor/blob/master/storage/elasticsearch/elasticsearch.go) made the same way.

File: plugins/outputs/elasticsearch/elasticsearch.go (hunk 10 above)
Comment (amandahla) [inline plugins/outputs/elasticsearch/elasticsearch.go]: Changed in commit:
https://github.com/influxdata/telegraf/pull/1875/commits/862f728635110c159444d2a88bdf3cdf9ae47edc

File: plugins/outputs/elasticsearch/elasticsearch.go (hunk 9 above)
Comment (amandahla) [inline plugins/outputs/elasticsearch/elasticsearch.go]: Changed in commit:
https://github.com/influxdata/telegraf/pull/1875/commits/862f728635110c159444d2a88bdf3cdf9ae47edc

**Comment by amandahla** [comment]
I made a commit with the requested changes. Only the test that you asked I'll take a little bit longer.

**Comment by amandahla** [comment]
I've written the test. Hope that I understood what you've asked for. :)

File: plugins/outputs/elasticsearch/elasticsearch_test.go (hunk 5 above)
Comment (amandahla) [inline plugins/outputs/elasticsearch/elasticsearch_test.go]: Changed in commit:
862f728

**Comment by rikatz** [comment]
Shoutout to @sparrc ! Is anything missing? 

This is a feature I really need :smile: And I think it would be useful to everyone.

Can I help @amandahla on something, so this PR may be accepted?

Thanks for your time !

**Comment by amandahla** [comment]
Hello. I changed the title because I tested with Elastic 5 Beta only and its different from the released one. I'm waiting a new version for Elastic now (https://github.com/olivere/elastic/issues/307).

**Comment by burdiyan** [comment]
Elastic search 5 is released. It would be great to have this output plugin!

**Comment by amandahla** [comment]
Hello @sparrc! Is there anything missing? Do I need to review anything else? Hope that all the changes that I did are ready for the PR.

**Comment by ajaybhatnagar** [comment]
I tried to use this telegraf module with elasticsearch as destination for data collected. Data collected for system metrics gets shipped OK and gets loaded but any data available thru jolokia or cassandra input streams never gets shipped. If check locally, the data is fetched.

wget http://localhost:8778/jolokia/read/wget http://localhost:8778/jolokia/read/org.apache.cassandra.metrics:type=Cache,scope=KeyCache,name=Capacity

gets following output:
{"request":{"mbean":"org.apache.cassandra.metrics:name=Capacity,scope=KeyCache,type=Cache","type":"read"},"value":{"Value":209715200},"timestamp":1479937498,"status":200}

Even index created has no field 
Config in telegraf.conf
 [[inputs.jolokia]]
context = "/jolokia/read"
   [[inputs.jolokia.servers]]
     name = "localhost"
     host = "127.0.0.1"
     port = "8778"
   [[inputs.jolokia.metrics]]
     name = "heap_memory_usage"
     mbean  = "java.lang:type=Memory"
     attribute = "HeapMemoryUsage"
   [[inputs.jolokia.metrics]]
     name = "cassandraCache"
     mbean  = "org.apache.cassandra.metrics:type=Cache"
     attribute = "KeyCache,Entries"

Running test shows following:
* Plugin: inputs.jolokia, Collection 1
Error handling response: Response from url "" has status code 404 (Not Found), expected 200 (OK)
Error handling response: Response from url "" has status code 404 (Not Found), expected 200 (OK)
Error handling response: Response from url "" has status code 404 (Not Found), expected 200 (OK)
Error handling response: Response from url "" has status code 404 (Not Found), expected 200 (OK)

**Comment by sparrc** [comment]
@ajaybhatnagar this is not the correct place for reporting issues with input plugins.

Please send general questions to the [influxdb@googlegroups.com](mailto:influxdb@googlegroups.com) mailing list or the #influxdb channel on [Gopher Slack](https://www.influxdata.com/announcing-the-influxdb-slack-channel/).

**Comment by megastef** [comment]
When will this be merged? Do you support _bulk inserts?

**Comment by sparrc** [comment]
I'd still like to see much better unit tests

File: README.md [hunk 11]
```diff
@@ -213,6 +213,46 @@ configuration options.
 
 Telegraf can also collect metrics via the following service plugins:
 
+* [http_listener](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/http_listener)
```
Comment (sparrc) [inline README.md]: these readme conflicts need to be fixed

File: plugins/outputs/elasticsearch/elasticsearch.go [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+	NumberOfShards   int
+	NumberOfReplicas int
+	Client           *elastic.Client
+	Version          string
+}
+
+var sampleConfig = `
+  server_host = "http://10.10.10.10:9200" # required.
+  index_name = "test" # required.
+  # regex allowed on index_name:
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch.go]: these are not regexes..? and can you provide examples for how they would be used?

File: plugins/outputs/elasticsearch/elasticsearch.go [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+	HealthCheck      bool
+	NumberOfShards   int
+	NumberOfReplicas int
+	Client           *elastic.Client
+	Version          string
+}
+
+var sampleConfig = `
+  server_host = "http://10.10.10.10:9200" # required.
+  index_name = "test" # required.
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch.go]: most of these config options seem to be elastic config options, can you provide links to documentation?

File: plugins/outputs/elasticsearch/elasticsearch.go [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+  # regex allowed on index_name: 
+  # %Y - year (2016)
+  # %y - last two digits of year (00..99)
+  # %m - month (01..12)
+  # %d - day of month (e.g., 01)
+  # %H - hour (00..23)
+  enable_sniffer = false
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch.go]: these options all need better documentation

File: plugins/outputs/elasticsearch/elasticsearch.go [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+		a.IndexName = strings.Replace(a.IndexName, "%H", fmt.Sprintf("%0.2d", time.Now().Hour()), -1)
+	}
+
+	exists, errExists := a.Client.IndexExists(a.IndexName).Do()
+
+	if errExists != nil {
+		return fmt.Errorf("FAILED to check if Elasticsearch index %s exists : %s\n", a.IndexName, errExists)
+	}
+
+	if !exists {
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/elasticsearch.go]: Do we need to do this? could we instead just require that users create an index before using this plugin?

File: plugins/outputs/elasticsearch/README.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+In this case, dots will be replaced with "_".
+
+### Configuration:
+
+```toml
+# Configuration for Elasticsearch to send metrics to
+[[outputs.elasticsearch]]
+  ## The full HTTP endpoint URL for your Elasticsearch. # required
+  server_host = "http://10.10.10.10:9200"
+  ## The target index for metrics # required
```
Comment (sparrc) [inline plugins/outputs/elasticsearch/README.md]: can you explain more about this auto-index-creating system you've come up with? why is it necessary? is it something the general user would need? Is there a precedent from another product?

also, if it needs to stay, you need to explain more about the template being used. When do the timestamps get evaluated? when does the index get created?, etc...

**Comment by ajaybhatnagar** [comment]
Is there a plan to integrate this output plugin with standard Telegraf builds and current version ?

**Comment by sparrc** [comment]
@ajaybhatnagar at the moment this PR still needs more work.

@amandahla are you able to continue working on this? I have a few other general questions:

1. Elasticsearch 5.x: Shouldn't we be using the elastic 5.x client library? What version(s) of elastic will this support? IMO we should only support 5.x
2. Does this plugin perform bulk inserts?

**Comment by ajaybhatnagar** [comment]
One more suggestion:
It should be possible to add custom Elasticsearch query thru configuration file in addition to standard metrics captured.

**Comment by sparrc** [comment]
@ajaybhatnagar why would you need to query elasticsearch? can you provide a detailed explanation and use-case?

**Comment by ajaybhatnagar** [comment]
There is need to  monitor critical indexes which grow very rapidly and each shard has a max number of documents. So one need to monitor when is the time create a new index or add more shards.

**Comment by lpic10** [comment]
I am volunteer to finish this, if @amandahla is ok with that I can push a new PR in the coming days.

IMO it should support only Elasticsearch 5

**Comment by berglh** [comment]
You can't add more shards other than replicas. Shards can be quite large, I
have shards at 20 GB and still working ok. With enough RAM. My
understanding is that ES 5 can handle index rotation autonomously now. You
don't need to monitor ES at all. The SysOp should know the limits and use
daily, monthly or yearly indices as required. This is adding unnecessary
complexity to what should be a lightweight metrics shipper to ES. The
output plugin should perform bulk writes using the ES bulk API.

**Comment by sparrc** [comment]
agree with @berglh, feel free to work on it @lpic10 work but do NOT add functionality for querying/modifying shard sizes and rotations.

**Comment by amandahla** [comment]
Hello, 
@ajaybhatnagar @sparrc I'm sorry but I can't continue working on this PR.
@lpic10 please feel free to work on this, I really appreciate.

**Comment by amandahla** [comment]
@lpic10 One question: Should I leave this PR open?

**Comment by lpic10** [comment]
You can close this, I will send a new PR when ready

**Comment by sparrc** [comment]
thanks all

**Comment by lmx1989219** [comment]
@amandahla
i like this pr,but why not go on this feature?

**Comment by amandahla** [comment]
@lmx1989219 You can see this feature in @lpic10 PR https://github.com/influxdata/telegraf/pull/2332

**Comment by fabMrc** [comment]
Hi ES 2.x compatibility was included in this PR , is it available or only 5.x is supported ?

**Comment by sparrc** [comment]
@fabMrc this PR was not merged, only 5.x is supported in the merged PR: https://github.com/influxdata/telegraf/pull/2332
